homeassistant/components/homekit_controller/connection.py

```diff
@@ -306,6 +315,22 @@ def _add_new_entities_for_accessory(self, handlers):
                     self.entities.append((accessory.aid, None))
                     break
 
+    def add_char_factory(self, add_entities_cb):
+        """Add a callback to run when discovering new entities for accessories."""
+        self.char_factories.append(add_entities_cb)
+        self._add_new_entities_for_char([add_entities_cb])
+
+    def _add_new_entities_for_char(self, handlers):
+        for accessory in self.entity_map.accessories:
+            for service in accessory.services:
+                for char in service.characteristics:
+                    for handler in handlers:
+                        if (accessory.aid, service.iid, char.iid) in self.entities:
+                            continue
+                        if handler(char):
+                            self.entities.append((accessory.aid, service.iid, char.iid))
+                            break
```

Comment:
This is painful, but I don't have a better way.

Reply:
Agreed, on both counts. I do have some cleanups planned around here that will help a bit.